### PR TITLE
Add CircleCI Cloud Config

### PR DESCRIPTION
**Pull request by briemcnally**
**What does this PR do?**

- Adds Circle CI Config for contractors that will run `yarn install`, `yarn lint`, and `yarn test`. 
- If the contributor makes changes to the `.buildkite` config file it will fail 
- If the contributor makes changes to the `yarn.lock`  file it will fail 

- **Will add the `.circleci/config.yml` to the shell script to fail if contributors make changes to that file. For the purposes of this PR if will make the Cloud CI fail, so would like a review prior to adding that change. It will currently fail if changes are made to `.buildkite` or `yarn.lock`.**

**Are there breaking changes in this PR?**
No. 

**Any background context you want to provide?**
This Circle CI Cloud Workflow will be used for Contractors to be run on Forked PRs. Internal CI Workflows for PRs will be run on Buildkite. Publishing to NPM will be handled by Buildkite. 

**Is there parity with the server-side/android/iOS integration components (if applicable)?**
N/A 

**Does this require a new integration setting? If so, please explain how the new setting works**
No

**Links to helpful docs and other external resources**
N/A

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            #Check the diff on the commits
+            git diff --name-only $COMMITS | cut -d/ -f1 | sort -u > projects
+            echo -e "Modified directories:\n`cat projects`\n"
+
+            # If modified directories contain yarn, or buildkite directories
+            buildall=0
+            for project in `cat projects`; do
+              if [[ ${project} =~ "yarn" || ${project} =~ ".buildkite" ]]; then
+                buildall=1
+                echo -e "Warning. Dependencies or internal configs have change detected."
```
Comment (CarlosMecha) [inline .circleci/config.yml]: This is AMAZING

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  lint:
+    docker:
+        - image: circleci/node:8-browsers
+    steps:
+      - attach_workspace: { at: . }
+      - restore_cache:
+          key: deps-{{ checksum "yarn.lock" }}
+      - run:
+          name: Run Linter (not enforcing for now)
+          command: yarn lint
```
Comment (CarlosMecha) [inline .circleci/config.yml]: Would be possible to run checkout, the check step, install, linter and tests in the same job? Attaching and detaching workspaces it's extremely slow with CI and I don't think we get much running all this in parallel.

File: .circleci/config.yml [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+jobs:
+  setup:
+    docker:
+      - image: circleci/node:8-browsers
+    steps:
+      - checkout
+      - run:
+          name: Determine which projects have changed and trigger the builds
+          command: |
+            ## Indentify head of master
```
Comment (omegabytes) [inline .circleci/config.yml]: Identify

File: .circleci/config.yml (hunk 2 above)
Comment (briemcnally) [reply .circleci/config.yml]: @CarlosMecha Just updated so it all runs under one job `build_and_test`.

File: .circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,55 @@
+version: 2
+jobs:
+  build_and_test:
```
Comment (CarlosMecha) [inline .circleci/config.yml]: Last nitpick, could we call this step `build`? Due to legacy naming, the default job on CircleCI is called `build`, and if you use their CLI for local testing, that's the only job that is able to run.

**Comment by briemcnally** [comment]
Note the `.circleci` config will now be failing since I have added `.circleci` to list of directories for the CI to fail on if it is modified. This is necessary to add since in the future we don't want contractors to edit the `.circleci` config but I needed to update it to migrate to the Circle CI Cloud config. I will use admin privileges to squash and merge this PR once buildkite passes. @CarlosMecha  @omegabytes

**Pull request merged**